### ISSUE TITLE
Model distillation leads to horrendous performance

### ISSUE THREAD

**Issue by lnatspacy** [UKPLab/sentence-transformers#1505]
Hi, 

Thank you for all your work. This is a fantastic library!
I trained a model using Online Contrastive Loss and it's working very well. 
It is however very slow on CPU and hence I wanted to reduce it's size using model distillation. I adapted your script slightly, removing the nli and stb datasets. 
Unfortunately, after training the performance is terrible and I don't really understand why. As a simple first evaluation I'm checking the cosine simillarity of some positive and negative samples and there is virtually no difference between positives and negatives. 

Here's the script I'm using: 

```
from torch.utils.data import DataLoader
from sentence_transformers import models, losses, evaluation
from sentence_transformers import LoggingHandler, SentenceTransformer
from sentence_transformers.datasets import ParallelSentencesDataset
import logging
import csv
from sklearn.decomposition import PCA
import torch


#### Just some code to print debug information to stdout
logging.basicConfig(format='%(asctime)s - %(message)s',
                    datefmt='%Y-%m-%d %H:%M:%S',
                    level=logging.INFO,
                    handlers=[LoggingHandler()])
#### /print debug information to stdout


# Teacher Model: Model we want to distill to a smaller model
teacher_model_name = 'my-teacher-model'
teacher_model = SentenceTransformer(teacher_model_name)

output_path = "my-teacher-model-distilled"

use_layer_reduction = True

#There are two options to create a light and fast student model:
if use_layer_reduction:
    # 1) Create a smaller student model by using only some of the teacher layers
    student_model = SentenceTransformer(teacher_model_name)

    # Get the transformer model
    auto_model = student_model._first_module().auto_model

    # Which layers to keep from the teacher model. We equally spread the layers to keep over the original teacher
    #layers_to_keep = [5]
    #layers_to_keep = [3, 7]
    #layers_to_keep = [3, 7, 11]
    #layers_to_keep = [1, 4, 7, 10]          #Keep 4 layers from the teacher
    layers_to_keep = [0, 2, 4, 6, 8, 10]
    #layers_to_keep = [0, 1, 3, 4, 6, 7, 9, 10]

    logging.info("Remove layers from student. Only keep these layers: {}".format(layers_to_keep))
    new_layers = torch.nn.ModuleList([layer_module for i, layer_module in enumerate(auto_model.encoder.layer) if i in layers_to_keep])
    auto_model.encoder.layer = new_layers
    auto_model.config.num_hidden_layers = len(layers_to_keep)
else:
    # 2) The other option is to train a small model like TinyBERT to imitate the teacher.
    # You can find some small BERT models here: https://huggingface.co/nreimers
    word_embedding_model = models.Transformer('nreimers/TinyBERT_L-4_H-312_v2')
    pooling_model = models.Pooling(word_embedding_model.get_word_embedding_dimension())
    student_model = SentenceTransformer(modules=[word_embedding_model, pooling_model])


inference_batch_size = 64
train_batch_size = 64

train_sentences = []
dev_sentences = []

csv.field_size_limit(1000000)

logging.info("Collecting training data")
with open("../data/sample_pairs.csv", encoding="utf-8") as f:
    csv_reader = csv.reader(f)    
    for cnt, sample in enumerate(csv_reader):
        train_sentences.append(sample[0])

logging.info("Collecting dev data")
with open("../data/dev_sample_pairs.csv", encoding="utf-8") as f:
    csv_reader = csv.reader(f)    
    for cnt, sample in enumerate(csv_reader):
        dev_sentences.append(sample[0])

train_sentences = train_sentences
dev_sentences = dev_sentences

# Student model has fewer dimensions. Compute PCA for the teacher to reduce the dimensions
if student_model.get_sentence_embedding_dimension() < teacher_model.get_sentence_embedding_dimension():
    logging.info("Student model has fewer dimensions than the teacher. Compute PCA for down projection")
    pca_sentences = train_sentences
    pca_embeddings = teacher_model.encode(pca_sentences, convert_to_numpy=True)
    pca = PCA(n_components=student_model.get_sentence_embedding_dimension())
    pca.fit(pca_embeddings)

    #Add Dense layer to teacher that projects the embeddings down to the student embedding size
    dense = models.Dense(in_features=teacher_model.get_sentence_embedding_dimension(), out_features=student_model.get_sentence_embedding_dimension(), bias=False, activation_function=torch.nn.Identity())
    dense.linear.weight = torch.nn.Parameter(torch.tensor(pca.components_))
    teacher_model.add_module('dense', dense)

    #logging.info("Teacher Performance with {} dimensions:".format(teacher_model.get_sentence_embedding_dimension()))
    #dev_evaluator_sts(teacher_model)

# We train the student_model such that it creates sentence embeddings similar to the embeddings from the teacher_model
# For this, we need a large set of sentences. These sentences are embedded using the teacher model,
# and the student tries to mimic these embeddings. It is the same approach as used in: https://arxiv.org/abs/2004.09813
logging.info("Creating dataset")
train_data = ParallelSentencesDataset(student_model=student_model, teacher_model=teacher_model, batch_size=inference_batch_size, use_embedding_cache=False)
train_data.add_dataset([[sent] for sent in train_sentences], max_sentence_length=512)

logging.info("Creating dataloader")
train_dataloader = DataLoader(train_data, shuffle=True, batch_size=train_batch_size)
train_loss = losses.MSELoss(model=student_model)

# We create an evaluator, that measure the Mean Squared Error (MSE) between the teacher and the student embeddings
logging.info("Creating evaluator")
dev_evaluator_mse = evaluation.MSEEvaluator(dev_sentences, dev_sentences, teacher_model=teacher_model)

# Train the student model to imitate the teacher
logging.info("Starting to fit model")
student_model.fit(train_objectives=[(train_dataloader, train_loss)],
                  evaluator=dev_evaluator_mse,
                  epochs=2,
                  warmup_steps=1000,
                  evaluation_steps=7000,
                  output_path=output_path,
                  save_best_model=True,
                  optimizer_params={'lr': 1e-4, 'eps': 1e-6, 'correct_bias': False},
                  use_amp=True)
```
Any help would be much appreciated !

**Comment by SantiDianaClibrain** [comment]
Hi! Were you able to solve this problem?
